[WiP] Updating for more complex experimental designs using study_variable[n]-group solution

sneumann opened the pull request.

sneumann committed:
started mentioning complex designs in section 5.2

sneumann committed:
Mark study-variable[1-n]-factor as deprecated.

sneumann committed:
Move the study_variable[1-n]-description right after assay_refs

sneumann committed:
Adding study_variable[1-n]-group

philouail commented:
`study-variable` is a required field, I feel that would be the `group` should be too ? and even if only one group of study variable is reported ? that would just make things much clearer and i do not think it is an issue for writer as there will always be a "column name". what do you think ?

nilshoffmann commented:
group should have the Parameter type. This will allow both CV Parameters and user parameters possible!
[CV, CV:1212312, name, value] -> CV Parameter
[,,"sex",] -> User Parameter

nilshoffmann commented:
Can we check the proposed naming against the one used by MetaboLights, so that we try to avoid adding more confusion.

nilshoffmann commented:
Suggested redesign, introducing a new top level group for study_variables:
```
study_variable_group[1]    [,,sex,]
study_variable_group[1]-description    ASjklhajksd
study_variable_group[1]-type   nominal
(study_variable_group[1]-unit   [,,,])? // e.g. day, time, concentration, ...
...
study_variable[1]    [,,female,]
study_variable[1]-group_ref    study_variable_group[1]
study_variable[1]-assay_refs    assay[1]|assay[3]|assay[5]
```

jorainer commented:
I very much like the new format with an explicit `study_variable_group`. I think it would be great to get that implemented ASAP. 

for `study_variable_group[1]-type` it would be good to have a consistent set of supported/expected values. don't know if there is a CV for that. But values that would come to my mind are (and the respective data types I would encode/decode that from R):

`"nominal"` -> use `as.is` in R (e.g. number to `numeric`, text to `character`)
`"ordinal"` -> convert to R `factor` (?)
`"categorical"` -> convert to R `factor` - question is just how to define the base level. easiest would be to have them ordered alphabetically.
`"logical"` -> `logical`.

nilshoffmann commented:
Look for MetaboLights public studies examples to inform some use cases for study_variable_group.

sneumann commented:
> if MetaboLights calls this _format_ then I would go for that. Using the same/similar nomenclature will make it less confusing for users (and developers as well).

If we can pursuade MTBLS that `format` is the least intuitive name, and what'd be what statisticians use, I'd rather try to agree on that. Hope it is not too late for MTBLS to switch...

philouail committed:
Merge branch 'main' into fix/issue40

philouail committed:
update group variable

jorainer commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2125,17 +2157,93 @@ components:
+          description: A textual description of the study variable group.
+        type:
+          type: string
+          enum:
```
This does not map 1:1 to the [STATO](https://www.ebi.ac.uk/ols4/ontologies/stato/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FSTATO_0000258?lang=en) ontology, right? but I actually prefer these *simpler* choices - might be easier for users to define add them, because most will not be statisticians and might get confused by too specific/detailed choices - just my opinion.

philouail commented on specification_documents/mzTab_format_specification_2_1-M.adoc:
```diff
@@ -1126,26 +1228,11 @@ MTD study_variable-variation_function [MS, MS:1002885, standard error, ]
 [cols=","]
 |=======================================================================================================================================================================
-|*Description* |Additional parameters or factors, separated by bars, that are known about study variables allowing the capture of more complex, such as nested designs.
+|*Description* |Additional parameters or factors, separated by bars, that are known about study variables allowing the capture of more complex, such as nested designs. *This field is deprecated and will be removed in a future release*.
```
Do we want to remove it in this PR ?

sneumann replied:
Ah, I did recall correctly in my comment above. +1 for removal

nilshoffmann replied:
We checked the old examples and the new ones for and only LipidCompass files have been using this.

nilshoffmann replied:
So ok to remove!

nilshoffmann replied:
TODO: Add an explanation to the spec doc about the removal and how to now encode this information.

philouail replied:
I added info about the removal. and I would say a more detailed explanation on encoding can come in another PR?

philouail commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2125,17 +2157,93 @@ components:
+          description: >
+            The data type of the group variable values.
+
+            "nominal" - Categorical values without inherent ordering; in R use as.is() or character vectors.
```
Yes but idk if this is the place, and i should then do it with other language

jorainer replied:
not all study variables will be categorical. at the moment it's only categorical

jorainer replied:
This one should be removed. No "nominal", but a "continuous" instead.

sneumann replied:
Changed

jorainer commented:
I'm OK with the changes, but we must also add a type to reflect "continuous variables", i.e., just a numeric. this is currently not listed as type

sneumann commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2125,17 +2157,93 @@ components:
+        description:
+          type: string
+          description: A textual description of the study variable group.
+        type:
```
Add date in ISO format

nilshoffmann replied:
Please update the term ids to the format CV:ID (compact URI), e.g. `[STATO, STATO:0000228, ordinal variable,]`.

nilshoffmann replied:
Looks good after the latest updates!

sneumann reviewed (commented):
Hi, sorry for taking a while. I think the structure could be merged right away, but the actual enums will need another round

sneumann commented on specification_documents/mzTab_format_specification_2_1-M.adoc:
```diff
@@ -1098,6 +1100,106 @@ MTD study_variable[1]-assay_refs assay[1]\| assay[2]\| assay[3]
+
+[cols=",",]
+|=============================================================================
+|*Description* |The data type of the group variable, which determines how the values should be interpreted or encoded/decoded. This field is MANDATORY when study_variable_group is defined to ensure unambiguous interpretation. The following data types are supported:
```
Hm, unsure if we can make that MANDATORY. Does the producer software always know what type the "sample class" is ? Does it know whether it is ordered or not ? Software will have a text input field, and users put in strings. That's all the software knows. I'll change to Mandatory=False

sneumann commented on specification_documents/mzTab_format_specification_2_1-M.adoc:
```diff
@@ -1098,6 +1100,106 @@ MTD study_variable[1]-assay_refs assay[1]\| assay[2]\| assay[3]
+
+Tools and parsers implementing mzTab-M SHOULD apply the appropriate data type interpretation when reading study_variable_group data and constructing analysis data structures (e.g., data frames, matrices, or tables).
+|*Type* |String (constrained)
+|*Mandatory* |True (if study_variable_group is defined)
```
Here and elsewhere: I think study_variable_group should be mandatory, so dropping that "if ... available"

sneumann committed:
Mostly study_variable_group work and smaller fixes

The majority of this commit is about the study_variable_group
- Adding an introductory secion 5.3. Recommendations for reporting multifactorial experimental designs
- Making the study_variable_group mandatory, to simplify parsing and processing
- Clarifying that study_variable_group[1:n]-type is a statistical type, referencing STATO
- Adding a study_variable_group[1:n]-datatype

Other changes include:
- Adding a section 4.1. Relationship to previous mzTab versions and a new subsection 4.1.2 4.1.2. Relationship to mzTab 2.0.0
- Fixing some adoc issues

sneumann committed:
Change table column relative widths for better layout

jorainer commented on schema/mzTab_m_openapi.yml:
```diff
@@ -719,7 +720,9 @@ components:
             description: A textual description of the study variable group.
 
-            type: The data type of the group variable (nominal, ordinal, categorical, or logical).
+            type: The statistical type of the group variable (nominal, ordinal, categorical, or logical).
```
wait - that differs now from what was discussed below. IMHO we should support 3 STATO terms: [nominal](http://purl.obolibrary.org/obo/STATO_0000252), [ordinal](http://purl.obolibrary.org/obo/STATO_0000228) and [continuous](http://purl.obolibrary.org/obo/STATO_0000251). logical is a special case for categorical (2 levels) and categorical is actually *nominal*. So, it all boils down to these 3 types.

philouail replied:
idk i i have addressed your concern here

jorainer commented on schema/mzTab_m_openapi.yml:
```diff
@@ -719,7 +720,9 @@ components:
-            type: The data type of the group variable (nominal, ordinal, categorical, or logical).
+            type: The statistical type of the group variable (nominal, ordinal, categorical, or logical).
+
+            datatype: The data type of the group variable.
```
is this really needed? I think it's pretty implicit/self explaining that the *nominal* type would match to string/character, *ordinal* should be integer (or at least numeric) and *continuous* numeric/real. So, IMHO we don't need this additional description.

sneumann replied:
That's correct if we treat them only as labels. But while true/false can be strings, they can also be booleans. And dates can be both ordinal and continuous (that'd be a design choice). 
So, the statistical type does not necessarily imply the datatype, More specifically, one needs the datatype to perform calculations like 
```
dates <- c("2025-05-10", "2025-05-25")
as.Date(dates[2]) - as.Date(dates[1])
```

jorainer replied:
that's true. so, this field could be used to help researchers understand the *format* in which the variable is encoded (which makes sense in particular for date formats to reduce ambiguity).

Basically, I'm OK with having this *datatype* as **optional** field. I'm not happy if that was a mandatory field.

nilshoffmann replied:
Dates should be encoded to be compatible with ISO8601, as defined in the specification document, line 387. https://github.com/HUPO-PSI/mzTab-M/edit/main/specification_documents/mzTab_format_specification_2_1-M.adoc
At the moment, we only state encoding for dates, we should extend that to date + time.

nilshoffmann replied:
Producers of mzTab-M 2.1.0 **should** add a datatype to simplify interpretation by downstream consumers of the format.

philouail replied:
I have updated the field please tell me if you agree

jorainer commented on examples/example_study_variable_group.mztab:
```diff
@@ -10,18 +10,21 @@ COM Study Variable Groups Definition
 MTD study_variable_group[1]-description	Sex of the individual
-MTD study_variable_group[1]-type	nominal
+MTD study_variable_group[1]-type	[STATO, STATO_0000252, categorical variable]
+MTD study_variable_group[1]-datatype	string
```
I don't think we need this additional *-datatype* - see comment further below.

jorainer commented on schema/mzTab_2_1-M_mapping.xml:
```diff
@@ -130,6 +130,17 @@ Changelog:
+    <CvMappingRule id="study_variable_group_type_may" cvElementPath="/metadata/studyVariableGroup/type/type/@Parameter" requirementLevel="MAY"  scopePath="/metadata/studyVariableGroup" cvTermsCombinationLogic="OR">
+      <CvTerm termAccession="STATO:0000252" useTerm="yes" termName="categorical variable" isRepeatable="false" allowChildren="true" cvIdentifierRef="STATO"></CvTerm>
+      <CvTerm termAccession="STATO:0000251" useTerm="yes" termName="continuous variable" isRepeatable="false" allowChildren="true" cvIdentifierRef="STATO"></CvTerm>
+    </CvMappingRule>
```
The *ordinal variable* http://purl.obolibrary.org/obo/STATO_0000228 is missing here.

sneumann replied:
Ordinal is a child of categorial, so it is covered by `allowChildren`.
Or, in line of the original discussion, leave it with the three explicitly allowed terms and don't allow children. That would be easier to document and explain.

philouail replied:
I have switched to `allowChildren = false`

jorainer replied:
agree. keep it simple.

nilshoffmann replied:
These are baseline recommendations anyway. So anyone can enforce harder rules on themselves if they really want to :-)

jorainer commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2077,7 +2106,7 @@ components:
         description: A textual description of the study variable.
 
-        factors: Additional parameters or factors, separated by bars, that are known about study variables allowing the capture of more complex, such as nested designs.  
+        factors: Additional parameters or factors, separated by bars, that are known about study variables allowing the capture of more complex, such as nested designs.
```
I would drop that. If there is a *complex, nested design* then it might need human intervention anyway, so it could equally well be described in the *description* part. So I would be for *drop this* to make it also easier for the mzTab-M importers/exporters.

sneumann replied:
I agree that complex nested designs will be challenging. 
But I still like the idea of having cvParams as variable instances. => Where would we put them ? 
into the `study_variable[1]      [PATO, PATO:0000383,female,]` ?

sneumann replied:
I also changed some examples to have the cvParam in `study_variable[1]      [PATO, PATO:0000383,female,]` now.

jorainer replied:
I would still like to remove this part (i.e., *factors*). IMHO the mzTab-M (mandatory) fields should be parsed and automatically interpreted by the software. Anything more complicated (such as (complex) study design) needs anyway human intervention. and that would be easier with purely descriptive text.

nilshoffmann replied:
The cvParam can also be a user param if it is not cv controlled, e.g. `[,,female,]`. This would require a change in the data model type, which would be not backwards compatible (just as a warning that we need to take care of this in the parsers).

jorainer commented on specification_documents/mzTab_format_specification_2_1-M.adoc:
```diff
@@ -1164,24 +1177,55 @@ MTD study_variable_group[1]-description Sex of the individual
+* **continuous** – `[STATO, STATO_0000251, continuous variable]` is a variable that can take any value within a range on a continuous scale. Example: temperature, height, concentration, time.
 
-* **logical** – Boolean/logical values (true/false, yes/no).
+//#* **logical** – Boolean/logical values (true/false, yes/no).
```
I would drop this. as we discussed, logical is actually categorical by definition. and in any statistical test (at least in R) it will be automatically converted to that (i.e. a `factor`)

sneumann replied:
yes, that's why it's commented out. So next stop: get rid of it.

jorainer commented on schema/mzTab_m_openapi.yml:
this should be

```
- ordinal
- categorical
- continuous
```

no?

sneumann replied:
fixed

jorainer commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2125,17 +2157,93 @@ components:
+
+            "nominal" - Categorical values without inherent ordering; in R use as.is() or character vectors.
+
+            "ordinal" - Ordered categorical values; in R convert to ordered factor.
```
In R ordinal are usually (AFAIK) encoded as integers, not as `factors`? or at least I would do that. for the statistical test then it might need to be converted or the correct test used.

sneumann replied:
By default R factors are unordered: 
```
> f <- factor(c("a", "b")) ; f[1] < f[2]
[1] NA
Warning message:
In Ops.factor(f[1], f[2]) : ‘<’ not meaningful for factors
```
but can be instanciated ordered (via natural ordering of the labels, or explicit `levels=`):
```
> f <- factor(c("a", "b"), ordered=TRUE) ; f[1] < f[2]
[1] TRUE
```

jorainer commented on schema/mzTab_m_openapi.yml:
Here we would then need the `"continuous"` to support any numeric variable (age, BMI, ...)

sneumann replied:
Indeed. But also, we'd move to the cvParam `[STATO, STATO_0000251, continuous variable]`

sneumann replied:
Done

jorainer commented on specification_documents/mzTab_format_specification_2_1-M.adoc:
```diff
@@ -1164,24 +1177,55 @@ MTD study_variable_group[1]-description Sex of the individual
+....
+
+|=============================================================================
+[[study_variable_group1-n-datatype]]
```
I am not convinced by this new entity. To me it is redundant to the `study_variable_group[1-n]-type`. In the end, e.g. dates will also be encoded as categorical data types (or ordinal) for the statistical tests.

sneumann replied:
see discussion above in the openapi corner of this PR

jorainer reviewed (changes requested):
Thanks for the updates @sneumann and @philouail ! There are however still some discrepancies in the document. Also, I'm not convinced by the addition of `study_variable_group[1-n]-datatype` (i.e., **data**type in addition to a type field.)

sneumann commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2125,17 +2157,93 @@ components:
         Specification of assay.
 
-        (empty) name: A name for each assay, to serve as a list of the assays that MUST be reported in the following tables. 
+        (empty) name: A name for each assay, to serve as a list of the assays that MUST be reported in the following tables.
```
ok, got it. It means `assay[1]` and empty sub-property, i.e. nothing like `-description`.

sneumann committed:
fixed some study_variable_group[]-type places to use STATO

sneumann reviewed (commented):
Hi, answering some of Jo's great comments. Still a few tweaks needed. Yours, Steffen

jorainer commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2219,22 +2218,18 @@ components:
-            The data type of the group variable values.
-
-            "nominal" - Categorical values without inherent ordering; in R use as.is() or character vectors.
+            The statistical type of the group variable, which determines how the values should be interpreted in a statistical analysis context. The type MUST be a term from the STATO ontology, and SHOULD be one of the examples below.
```
This is **mandatory**, right?

jorainer commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2219,22 +2218,18 @@ components:
+            [STATO, STATO_0000251, continuous variable]` is a variable that can take any value within a range on a continuous scale.
+        datatype:
+          type: string
+          description: The datatype of the group variable, which determines how they can be encoded and parsed, and how the values could be handled in programming languages.
```
*An optional definition of the datatype for the group variable, ...*

jorainer reviewed (approved):
I'm in general OK with the proposed changes. Only things to discuss/change would be the that *datatype* for the study group variable should be optional and to remove the *factors: Additional parameters or factors, separated by ...* part.

nilshoffmann commented on specification_documents/mzTab_format_specification_2_1-M.adoc:
```diff
@@ -1098,6 +1100,106 @@ MTD study_variable[1]-assay_refs assay[1]\| assay[2]\| assay[3]
+==== study_variable_group[1-n]-type
+
+[cols=",",]
+|=============================================================================
```
To keep in sync with the original study variable spec for cases where no such data is being captured, the study_variable_group should have the following fallback (suggestion): 
For software that does not capture study variables, a single study variable group MUST be reported, linking to the single study variable. This single study variable group MUST have the identifier “undefined“.

philouail replied:
I have added this mention in the description above. Tell me if that is enough

nilshoffmann commented on specification_documents/mzTab_format_specification_2_1-M.adoc:
```diff
@@ -1205,7 +1203,7 @@ MTD study_variable_group[3]-type [STATO, STATO_0000251, continuous variable]
 [cols="20,80",]
 |=============================================================================
-|*Description* |The datatype of the group variable, which determines how they can be encoded and parsed in mzTab-M files, and how the values are handled in programming languages. Datatypes are specified as W3C XML Schema datatypes.
+|*Description* |The datatype of the group variable, which determines how they can be encoded and parsed in mzTab-M files, and how the values could be handled in programming languages. 
```
Dates and Datetimes **MUST** conform to ISO8601, as defined in the specification document.
For other formats, we should agree on a list?

philouail committed:
Merge branch 'main' into fix/issue40

philouail committed:
fix type and datatype, remove factors.

philouail commented on schema/mzTab_m_openapi.yml:
```diff
@@ -255,18 +255,18 @@ components:
+            * Small Molecule Feature (SMF) and the
+            * Small Molecule Evidence (SME).
+
+        The MTD and SML tables are mandatory, and for a file to contain any evidence about how molecules were quantified or identified by software, then all four tables must be present. The tables must follow the order MTD, SML, SMF and SME, with a blank line separating each table.
```
Yes you're right but i believe this should be part of a separate PR. and accompanied with a minimal example.

nilshoffmann replied:
Created #104 for continued discussion and proposals.

philouail committed:
fix CV:ID format

jorainer commented on schema/mzTab_m_openapi.yml:
```diff
@@ -708,21 +706,19 @@ components:
 
             description: A textual description of the study variable group.
 
             type: The statistical type of the group variable, which determines how the values should be interpreted in a statistical analysis context. The type MUST be a term from the STATO ontology, and SHOULD be one of [STATO, STATO_0000252, categorical variable], [STATO, STATO_0000228, ordinal variable], or [STATO, STATO_0000251, continuous variable].
```
could you please change here `_` to `:` , i.e. `STATO_...` to `STATO:...`

philouail replied:
ah shoot i thought i fixed all of them

philouail replied:
ah yes all good i looked at the wrong commit aha

jorainer reviewed (changes requested):
Agree with all changes, only please change the `STATO_...` to `STATO:...` @philouail

jorainer reviewed (approved):
Forget my previous review :)

nilshoffmann commented on schema/mzTab_m_openapi.yml:
```diff
@@ -707,9 +706,42 @@ components:
-            MTD	study_variable[1]-factors	[,,rapamycin dose,0.5mg]
-
             MTD	study_variable[2]	1 minute
+        study_variable_group:
```
Please move study_variable_group before the study_variable definition to reflect the logical order of definition when writing or reading the file.

nilshoffmann commented on schema/mzTab_m_openapi.yml:
```diff
@@ -2111,31 +2139,105 @@ components:
         description:
           type: string
           description: A free-form description of this study variable.
-        factors:
```
I think we might have an issue here when we want to encode different levels that are Parameter controlled. Here is an example that I currently have for LipidCompass:

```
MTD     study_variable[1]       Indian
MTD     study_variable[1]-description   Study population with mainly Indian ancestry
MTD     study_variable[1]-factors       [NCIT, NCIT:C16564, Ethnic Group, Indian]
MTD     study_variable[1]-assay_refs    ...
MTD     study_variable[2]       Chinese
MTD     study_variable[2]-description   Study population with mainly Chinese ancestry
MTD     study_variable[2]-factors       [NCIT, NCIT:C16564, Ethnic Group, Chinese]
MTD     study_variable[2]-assay_refs    ...
MTD     study_variable[3]       Malayan
MTD     study_variable[3]-description   Study population with mainly Malayan ancestry
MTD     study_variable[3]-factors       [NCIT, NCIT:C16564, Ethnic Group, Malay]
MTD     study_variable[3]-assay_refs    ...
```

Here, the factors are used to represent distinct CV terms. If I try to adapt this to the new `study_variable_groups`, I have the following structure:

```
MTD     study_variable_group[1]     [NCIT: NCIT:C16464, Ethnic Group,]
MTD     study_variable_group[1]-description    Ethnic background of study group
MTD     study_variable_group[1]-type	[STATO, STATO:0000252, categorical variable] //defines the type of values occuring in the study_variable[1-n]   value slot
MTD     study_variable_group[1]-datatype	xsd:string // maybe allow Param alternatively? This would be a mixed type that we haven not had before, so may be difficult to implement.
MTD     study_variable[1]       Indian // in this case, a string is ok, but should a Cv- or UserParam be possible as well?
MTD     study_variable[1]-description   Study population with mainly Indian ancestry
MTD     study_variable[1]-assay_refs    ...
MTD     study_variable[2]       Chinese
MTD     study_variable[2]-description   Study population with mainly Chinese ancestry
MTD     study_variable[2]-assay_refs    ...
MTD     study_variable[3]       Malayan
MTD     study_variable[3]-description   Study population with mainly Malayan ancestry
MTD     study_variable[3]-assay_refs    ...
```

For many use cases, this might be sufficient, unless the levels themselves represent CvParams.

jorainer replied:
but then - why not provide the CV term as either the `study_variable[3]-description` or as the value for `study_variable[1]`? or we could have an *additional* field for that, e.g. `study_variable[1]-cv_param`? My issue with the *factors* was that I did not understand what it meant from the documentation. I would be very explicit and have clearly (ideally self-explanatory) fields.

nilshoffmann replied:
In order to be slightly more backwards compatible,
`study_variable[1]-cv_param` would be ok, even though we do not have a precedent for this at the moment.
Setting the parameter on the study_variable would be cleaner solution and would be in line with many other similar objects in the model, so I would suggest we use this option:

```
MTD     study_variable[1]       [,,Ethnic Group, Indian]
```

```
MTD     study_variable[1]       [NCIT,NCIT:C16464, Ethnic Group, Indian]
```

This may be a bit repetitive when `study_variable_group` is being used, but is also explicit and would allow an additional cross-check.

jorainer replied:
so, you propose to change the `study_variable` to a param, right? that would be OK for me. in our use-cases most of the time it would then be a user param in the form of `[,,<column name>,<value>]` where `<column name>` would also be the name of the `study_variable_group`.

nilshoffmann replied:
Ok, so let's change the type of study_variable value to cvParam.

nilshoffmann commented on schema/mzTab_m_openapi.yml:
```diff
@@ -402,6 +402,7 @@ components:
         - ms_run
         - assay
         - study_variable
+        - study_variable_group
```
Please swap study_variable and study_variable_group

nilshoffmann commented on examples/example_study_variable_group.mztab:
```diff
@@ -0,0 +1,129 @@
+MTD study_variable_group[1]-datatype	xsd:string
+
+COM timepoint group - ordinal type (ordered categories) with unit
+MTD study_variable_group[2]	[,,timepoint,]
```
study_variable_group[2] and study_variable_group[3] are never referenced by a study_variable. 
If I don't misinterpret design, we are missing a relation here? 
We may need to group-ref multiple study_variable_groups, like 

```
...
MTD study_variable[1]-group_ref	study_variable_group[1]|study_variable_group[2]|study_variable_group[3]
...
```

However, how do we then read out the individual values? We may need to allow linkage on the sample level, if I am not mistaken here? Or how would I read back the treatment time point for a specific timepoint? At the moment, the study_variable name would need to be parsed (and conform to a specific convention).

nilshoffmann replied:
Scenario 1) we **do not** care about the individual values -> we should still group_ref(s) to all relevant groups.
Scenario 2) we **do** care about the individual values -> these need to occur on the sample level, via the assay_ref indirection. But here, we do not have a direct way to map this. We may already cover the information in one of the specific, parameter controlled attributes of Sample, however we may not want to change the currrent baseline behavior. Thus, we should consider adding a new attribute to Sample

```
MTD    sample[1]-study_variable_group[1]-value    [,,sex, Male]
MTD    sample[1]-study_variable_group[2]-value    [,,timepoint, 0]
MTD    sample[1]-study_variable_group[3]-value    [,,genotype, WT]
```

This would allow to encode both "simple" values, as well as (CV)parameters.

jorainer replied:
Maybe I miss something - but how we are encoding the things at the moment is:

given: data from 4 individuals, 3 covariates available:

| sex | BMI | T2D |
|----|----|----|
| M | 31.2 | TRUE |
| F | 26.4 | FALSE |
| F | 29.4 | TRUE |
| M | 29.5 | FALSE |

we would encode this into 3 `study_variable_group`s, one for each column (covariate). Then, we would have 2 study variables (M, F) assigned to the first group, 4 study variables to the second and 2 to the third; with the assay refs than pointing to the individual rows. So, the references between them would be encoded like:

```
study_variable_group[1]    sex
study_variable_group[2]    BMI
study_variable_group[3]    T2D
study_variable[1]    M
study_variable[1]-group_refs    study_variable_group[1]
study_variable[1]-assay_refs    assay[1]|assay[4]
study_variable[2]    F
study_variable[2]-group_refs    study_variable_group[1]
study_variable[2]-assay_refs    assay[2]|assay[3]
study_variable[3]    31.2
study_variable[3]-group_refs    study_variable_group[2]
study_variable[3]-assay_refs    assay[1]
study_variable[4]    26.4
study_variable[4]-group_refs    study_variable_group[2]
study_variable[4]-assay_refs    assay[2]
study_variable[5]    29.4
study_variable[5]-group_refs    study_variable_group[2]
study_variable[5]-assay_refs    assay[3]
study_variable[6]    26.5
study_variable[6]-group_refs    study_variable_group[2]
study_variable[6]-assay_refs    assay[4]
study_variable[7]    TRUE
study_variable[7]-group_refs    study_variable_group[3]
study_variable[7]-assay_refs    assay[1]|assay[3]
study_variable[8]    FALSE
study_variable[8]-group_refs    study_variable_group[4]
study_variable[8]-assay_refs    assay[2]|assay[4]
```

IMHO this should be enough to cover [any/most] experimental designs... I don't see an advantage of assigning a study group variable **value** also to a sample. then we encode the same thing twice. and even worse if we make that optional, because then we need in the importer/exporter also consider that - and we **have** to cross-check that the information is actually correct.

I would opt to keep it simple. the simpler the standard, the easier it might get picked up and actually used...

nilshoffmann replied:
Thansk for the example. I agree, this looks good.
Please note that the example [examples/example_study_variable_group.mztab](https://github.com/HUPO-PSI/mzTab-M/pull/77/files/a62227cc865e8ba5c5d803251ae944b7cfb440b7#diff-00ed47e9a3f7006a017bda3ebf7681a4c8da01fd6d04f90ebcadce3bc289d259) still needs to be fixed.

nilshoffmann reviewed (changes requested):
Most points are ready to merge. Please resolve the point about the order of attributes and definitions for study_variable_group in the api description and then we should be good for now. I have created two new issues to track some points we should discuss.
